### PR TITLE
[Issue 2334] Fix state transition committee cache

### LINKED ISSUE #2334: State Transition error - Invalid attestation: Signature is invalid
Seen this a handful of times in the logs, most recently today. If it's due to receiving an invalid block from a peer, then there's not much an operator can do about this, so it is probably better to move it to DEBUG level. But it might also be due to something internal that needs to be fixed.

```
2020-07-09 00:26:41.512+00:00 | ValidatorApiChannel-0 | INFO  | teku-validator-log | ^[[34mValidator   *** Published block        Count: 1, Slot: 68383, Root: 6017b1..5c80^[[0m
2020-07-09 00:26:41.614+00:00 | events-420 | WARN  | StateTransition | State Transition error
tech.pegasys.teku.core.exceptions.BlockProcessingException: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
        at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:88) [teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.ForkChoiceUtil.on_block(ForkChoiceUtil.java:223) [teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.onBlock(ForkChoice.java:77) [teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.blockimport.BlockImporter.importBlock(BlockImporter.java:70) [teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.blockimport.BlockImporter.onBlockProposed(BlockImporter.java:100) [teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at jdk.internal.reflect.GeneratedMethodAccessor75.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:567) ~[?:?]
        at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:87) [guava-28.1-jre.jar:?]
        at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:144) [guava-28.1-jre.jar:?]
        at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:72) [guava-28.1-jre.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
        at tech.pegasys.teku.core.BlockProcessorUtil.verify_attestations(BlockProcessorUtil.java:399) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.SimpleBlockValidator.validatePreState(SimpleBlockValidator.java:78) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:43) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.BlockValidator.validate(BlockValidator.java:88) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:85) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        ... 13 more
2020-07-09 00:26:41.619+00:00 | events-420 | ERROR | BlockImporter | Failed to import proposed block for reason + FAILED_STATE_TRANSITION: tech.pegasys.teku.statetransition.events.block.ProposedBlockEvent@4037b97d
tech.pegasys.teku.core.StateTransitionException: tech.pegasys.teku.core.exceptions.BlockProcessingException: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
        at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:97) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.ForkChoiceUtil.on_block(ForkChoiceUtil.java:223) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.onBlock(ForkChoice.java:77) ~[teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.blockimport.BlockImporter.importBlock(BlockImporter.java:70) ~[teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.statetransition.blockimport.BlockImporter.onBlockProposed(BlockImporter.java:100) [teku-ethereum-statetransition-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at jdk.internal.reflect.GeneratedMethodAccessor75.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:567) ~[?:?]
        at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:87) [guava-28.1-jre.jar:?]
        at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:144) [guava-28.1-jre.jar:?]
        at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:72) [guava-28.1-jre.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: tech.pegasys.teku.core.exceptions.BlockProcessingException: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
        at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:88) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        ... 13 more
Caused by: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
        at tech.pegasys.teku.core.BlockProcessorUtil.verify_attestations(BlockProcessorUtil.java:399) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.SimpleBlockValidator.validatePreState(SimpleBlockValidator.java:78) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:43) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.blockvalidator.BlockValidator.validate(BlockValidator.java:88) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:85) ~[teku-ethereum-core-0.12.1-SNAPSHOT.jar:0.12.1-dev-3e9bc8a3]
        ... 13 more
```

```
teku/v0.12.1-dev-3e9bc8a3/linux-x86_64/-privatebuild-openjdk64bitservervm-java-13
```

### LINKED ISSUE #2446: Dropped out of sync on Altona
### Description
My local node has dropped out of sync on Altona.  AWS nodes we run have continued tracking head so this is not a straight consensus split - more likely our sync code getting stuck somewhere.

I have a lot of

```
ku_1           | 2020-07-26 23:54:06.041 | p2p-async-4               | WARN  | PeerSync                       | Failed to import block from peer (err: FAILED_STATE_TRANSITION) SignedBeaconBlock{message=BeaconBlock{root=0x6f3dd4fdc0b4a30e911d607c8810f4a6e67253be2925fb3ea51efdcefa4aedc3, slot=190953, proposer_index=4766, parent_root=0x961416590b72bee69ac3de2b5f368be08e7e3561ea1b5ca8319bb8ba20639080, state_root=0xd8671292df6bc47bddab7aceb1c90adeebae72778bbf1523f87edc3e7b8c473a, body=0x773fd31666f17bfc8368599b6affaf3dd28e205c3e096390b82bec62c9271da0}, signature=0x8bac93e329cdfe5c48b630df82baf168682038a1bb424e763f6f884b5a80574a1e201e38daa0259771a4310e4564c4f00eee23bd2d0ac6a2be98d1147d20a3002923260337149e24d405ba1c9610c4c1b610fd55f778d1a2ec69eae7d6254527}: Eth2Peer{id=16Uiu2HAmGuSCvqbtGYJtUUwwrcC697HoDmGEwTXBGNbi4DpEYCip, remoteStatus=Optional[PeerStatus{forkDigest=0xfdca39b0, finalizedRoot=0x028460d5ed683b82f43b7f353a6131a1bd66d2231af79b2c96ef4eec3ca741e3, finalizedEpoch=6170, headRoot=0x3130976da9ce87d3117fff850215218ffd1d96e99e580b728717a4487a89737a, headSlot=197818}]}
teku_1           | 2020-07-26 23:54:17.210 | TimeTickChannel-0         | INFO  | teku-event-log                 | Sync Event  *** Current slot: 197821, Head slot: 190950, Connected peers: 5
teku_1           | 2020-07-26 23:54:29.210 | TimeTickChannel-0         | INFO  | teku-event-log                 | Sync Event  *** Current slot: 197822, Head slot: 190950, Connected peers: 7
teku_1           | 2020-07-26 23:54:35.574 | p2p-async-2               | WARN  | StateTransition                | State Transition error
teku_1           | tech.pegasys.teku.core.exceptions.BlockProcessingException: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
teku_1           | 	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:109) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.core.ForkChoiceUtil.on_block(ForkChoiceUtil.java:229) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.onBlock(ForkChoice.java:79) ~[teku-ethereum-statetransition-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.statetransition.blockimport.BlockImporter.lambda$importBlock$0(BlockImporter.java:73) ~[teku-ethereum-statetransition-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680) ~[?:?]
teku_1           | 	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658) ~[?:?]
teku_1           | 	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2158) ~[?:?]
teku_1           | 	at tech.pegasys.teku.infrastructure.async.SafeFuture.thenApply(SafeFuture.java:309) ~[teku-infrastructure-async-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.statetransition.blockimport.BlockImporter.importBlock(BlockImporter.java:71) ~[teku-ethereum-statetransition-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.sync.PeerSync.blockResponseListener(PeerSync.java:216) ~[teku-sync-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.sync.PeerSyncBlockRequest.onResponse(PeerSyncBlockRequest.java:41) ~[teku-sync-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.sync.PeerSyncBlockRequest.onResponse(PeerSyncBlockRequest.java:22) ~[teku-sync-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamImpl.respond(ResponseStreamImpl.java:67) ~[teku-networking-eth2-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.networking.eth2.rpc.core.AsyncResponseProcessor.lambda$processNextResponse$0(AsyncResponseProcessor.java:108) ~[teku-networking-eth2-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.infrastructure.async.SafeFuture.ofComposed(SafeFuture.java:85) ~[teku-infrastructure-async-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:119) ~[teku-infrastructure-async-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
teku_1           | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
teku_1           | 	at java.lang.Thread.run(Thread.java:832) [?:?]
teku_1           | Caused by: tech.pegasys.teku.core.exceptions.BlockProcessingException: Invalid attestation: Signature is invalid
teku_1           | 	at tech.pegasys.teku.core.BlockProcessorUtil.verify_attestations(BlockProcessorUtil.java:380) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.SimpleBlockValidator.validatePreState(SimpleBlockValidator.java:78) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:43) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BlockValidator.validate(BlockValidator.java:88) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:106) ~[teku-ethereum-core-0.12.2-SNAPSHOT.jar:0.12.2-dev-95530612]
teku_1           | 	... 18 more
```

Complete logs since last restart:
[tekuLogs.txt](https://github.com/PegaSysEng/teku/files/4978996/tekuLogs.txt)

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
This PR fixes an issue where our state transition caches were being corrupted with invalid committee data.

I was able to reproduce this corrupted cache issue as follows:
* Start a single node network with 64 validators
* Let the network run for a bit and finalize some epochs
* Stop the node and let several epoch pass
* Restart the node, at which point we would get "Invalid attestation: Signature is invalid" errors

On restart, the node would calculate and schedule upcoming validator duties correctly.  However, when we went to produce our attestations, we would try to calculate committee information using an old state (from the head of the chain) without processing slots forward to the slot being queried.  The calculated committee information from the stale state would then end up being cached by our `TransitionCaches` object.  This caused subsequent queries to pull committee information to return invalid data.  This then caused us to view our produced attestations as invalid because the corrupted committee information was inconsistent with the produced attestation. 

The fix is to ensure that any state used to calculate committee information is in range of the slot being queried (the state must be from at least the previous epoch).  Checks have been added to `CommitteeUtil.get_beacon_committee()` to make sure we don't allow old states to be used for this calculation. `ChainDataProvider` and `ValidatorApiHandler` have been updated to run with states that have been processed up to an appropriate epoch before trying to calculate committee information.

## Fixed Issue(s)
Fixes #2334 
Fixes #2446

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java [hunk 1]
```diff
@@ -83,25 +81,22 @@ public GetForkResponse getForkInfo() {
...
-    final UnsignedLong slot = compute_start_slot_at_epoch(epoch);
-
-    // one epoch in future is available, beyond that cannot be calculated
-    if (slot.compareTo(
-            recentChainData.getBestSlot().plus(UnsignedLong.valueOf(Constants.SLOTS_PER_EPOCH)))
-        > 0) {
+
+    final UnsignedLong earliestQueryableSlot =
+        CommitteeUtil.getEarliestQueryableSlotForTargetEpoch(epoch);
+    if (recentChainData.getBestSlot().compareTo(earliestQueryableSlot) < 0) {
```
Comment (mbaxter) [inline data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:87]: This check means that if we have a span of time where no blocks are produced, we won't be able to query committee information for epochs ahead of the epoch after the latest block (from the REST API).  Is this right?

Comment (ajsutton) [reply data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:87]: I think we now progress the best slot even if the slots are empty so this should work out ok.  Worth testing to be sure though.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/CheckpointState.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          new BeaconBlockHeader(
+              latestHeader.getSlot(),
+              latestHeader.getProposer_index(),
+              latestHeader.getParent_root(),
+              stateRoot,
+              latestHeader.getBody_root());
+    }
+
+    return latestHeader;
+  }
```
Comment (ajsutton) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/CheckpointState.java]: This feels like a pretty expensive test to perform just to be defensive. I'm be tempted to omit it.

**Review by ajsutton — APPROVED** [review]
LGTM. Nice work.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/CheckpointState.java (hunk 2 above)
Comment (mbaxter) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/CheckpointState.java]: yeah - I think I got a little carried away trying to figure out how to validate this.  probably not worth including.

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java (hunk 1 above)
Comment (mbaxter) [reply data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:87]: The slot we pull is the slot from the head block, not the latest slot processed.  I think what we probably want to do here is return an empty result if we're syncing, or return an error if the slot is from the future. 🤔

Comment (ajsutton) [reply data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:87]: Hmm, you'er right but we do track the actual best slot - `updateBestBlock` is called with the nodeSlot even if the slot was empty to ensure we treat It as a reorg if we later import blocks at those slots.  So if `getBestSlot` (or possibly a new method) returned `chainHead.getSlot()` we'd get the current slot regardless of any empty slots.

**Pull request merged**